build_tools/cmake/iree_fetch_artifact.cmake

```diff
@@ -0,0 +1,53 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# iree_fetch_artifact()
+#
+# Download file from URL. NEVER Use this rule to download from untrusted
+# sources, it doesn't unpack the file safely.
+#
+# Parameters:
+# NAME: Name of target (see Note).
+# SOURCE_URL: Source URL to donwload the file.
+# OUTPUT: Path to the output file or directory to unpack.
+# UNPACK: When added, it will try to unpack the archive if supported.
+#
+# Note:
+# By default, it will create a target named ${_PACKAGE_NAME}_${_RULE_NAME}.
+function(iree_fetch_artifact)
+  cmake_parse_arguments(
+    _RULE
+    "UNPACK"
+    "NAME;SOURCE_URL;OUTPUT"
+    ""
+    ${ARGN}
+  )
+
+  set(_ARGS "${IREE_ROOT_DIR}/build_tools/scripts/download_file.py")
+  list(APPEND _ARGS "${_RULE_SOURCE_URL}")
+  list(APPEND _ARGS "-o")
+  list(APPEND _ARGS "${_RULE_OUTPUT}")
+
+  if(_RULE_UNPACK)
+    list(APPEND _ARGS "--unpack")
+  endif()
+
+  add_custom_command(
+    OUTPUT "${_RULE_OUTPUT}"
+    COMMAND
+      "${Python3_EXECUTABLE}"
+      ${_ARGS}
+    DEPENDS
+      "${IREE_ROOT_DIR}/build_tools/scripts/download_file.py"
+    COMMENT "Downloading ${_RULE_SOURCE_URL}"
+  )
```

Comment:
Can you add a comment here explaining why we don't use `file(DOWNLOAD`?
https://cmake.org/cmake/help/latest/command/file.html#transfer

* Similarly for `file(ARCHIVE_EXTRACT`: https://cmake.org/cmake/help/latest/command/file.html#archiving
* and possibly also `FetchContent`: https://cmake.org/cmake/help/latest/module/FetchContent.html

(all of this can be done without requiring Python, see e.g. https://www.scivision.dev/cmake-download-verify-file/)

Reply:
Also fine to add a TODO here to explore using the native `file()` functions, and keep this PR focused on the function refactoring.

Reply:
Added the TODO.

The python script unpacks the archive in a streaming manner, so it doesn't use double disk space. This is one of the advantages of using the script. Also we might want to implement a retry logic. I don't know if CMake supports that.